pkg/metering/writer.go

```diff
@@ -0,0 +1,209 @@
+		Prefix: meteringConfig.Prefix,
+		AWS: &storage.AWSConfig{
+			AssumeRoleARN: meteringConfig.RoleARN,
+		},
+	}
+	provider, err := storage.NewObjectStorageProvider(providerConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create storage provider for metering")
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	return &Writer{
+		id:       id,
+		ctx:      ctx,
+		cancel:   cancel,
+		provider: provider,
+		inner: meteringwriter.NewMeteringWriter(
+			provider,
+			config.DefaultConfig().WithLogger(zap.L()),
+		),
+		collectors: make(map[string]Collector),
+	}, nil
+}
+
+// Start starts the metering writer background loop
+func (mw *Writer) Start() {
+	if mw == nil {
+		return
+	}
+	mw.wg.Add(1)
+	go mw.meteringLoop()
+	log.Info("metering writer started")
+}
+
+// Stop stops the metering writer
+func (mw *Writer) Stop() {
+	if mw == nil {
+		return
+	}
+	if mw.cancel != nil {
+		mw.cancel()
+	}
+	mw.wg.Wait()
+	if mw.inner != nil {
+		mw.inner.Close()
+	}
+	log.Info("metering writer stopped")
+}
+
+// RegisterCollector registers a collector for a category. Overwrites existing if any.
+func (mw *Writer) RegisterCollector(collector Collector) {
+	if mw == nil {
+		return
+	}
+	mw.Lock()
+	defer mw.Unlock()
+	if mw.collectors == nil {
+		mw.collectors = make(map[string]Collector)
+	}
+	mw.collectors[collector.Category()] = collector
+}
+
+// meteringLoop runs the background loop to flush metering data periodically.
+func (mw *Writer) meteringLoop() {
+	defer logutil.LogPanic()
+	defer mw.wg.Done()
+
+	now := time.Now()
+	// Truncate to the nearest minute and add the interval to get the next flush time.
+	next := now.Truncate(meteringFlushInterval).Add(meteringFlushInterval)
+	log.Info("metering writer loop started",
+		zap.String("self-id", mw.id),
+		zap.Time("now", now),
+		zap.Time("next", next))
+	for {
+		select {
+		case <-mw.ctx.Done():
+			ts := next.Unix()
+			log.Info("context done received, flushing metering data one last time",
+				zap.Int64("timestamp", ts))
+			mw.flushMeteringData(ts)
+			log.Info("metering writer loop exits")
+			return
+		case <-time.After(next.Sub(now)):
+			mw.flushMeteringData(next.Unix())
+			now = time.Now()
+			next = next.Add(meteringFlushInterval)
+		}
+	}
+}
+
+// flushMeteringData flushes aggregated metering data to the underlying storage
+func (mw *Writer) flushMeteringData(ts int64) {
+	collectors := mw.getCollectors()
+	if len(collectors) == 0 {
+		return
+	}
+
+	for category, collector := range collectors {
+		records := collector.Flush()
+		if len(records) == 0 {
+			continue
+		}
+		// This should never happen in the real world, it's used to avoid panic during testing.
+		if mw.inner == nil {
+			continue
+		}
+		meteringData := &common.MeteringData{
+			SelfID:    mw.id,
+			Timestamp: ts,
+			Category:  category,
+			Data:      records,
+		}
+		recordCount := len(records)
+		// TODO: write with pagination if needed.
+		if err := mw.inner.Write(mw.ctx, meteringData); err != nil {
```

Comment:
The mw.ctx may be done in the last round, so it's useless to flush it after ctx.Done().